azurelinuxagent/common/utils/restutil.py

```diff
@@ -165,6 +169,93 @@ def _parse_url(url):
     return o.hostname, o.port, secure, rel_uri
 
 
+def is_valid_cidr(string_network):
+    """
+    Very simple check of the cidr format in no_proxy variable.
+    :rtype: bool
+    """
+    if string_network.count('/') == 1:
+        try:
+            mask = int(string_network.split('/')[1])
+        except ValueError:
+            return False
+
+        if mask < 1 or mask > 32:
+            return False
+
+        try:
+            socket.inet_aton(string_network.split('/')[0])
+        except socket.error:
+            return False
+    else:
+        return False
+    return True
+
+
+def address_in_network(ip, net):
+    """This function allows you to check if on IP belongs to a network subnet
+    Example: returns True if ip = 192.168.1.1 and net = 192.168.1.0/24
+             returns False if ip = 192.168.1.1 and net = 192.168.100.0/24
+    :rtype: bool
+    """
+    ipaddr = struct.unpack('=L', socket.inet_aton(ip))[0]
+    netaddr, bits = net.split('/')
+    netmask = struct.unpack('=L', socket.inet_aton(dotted_netmask(int(bits))))[0]
+    network = struct.unpack('=L', socket.inet_aton(netaddr))[0] & netmask
+    return (ipaddr & netmask) == (network & netmask)
+
+
+def dotted_netmask(mask):
+    """Converts mask from /xx format to xxx.xxx.xxx.xxx
+    Example: if mask is 24 function returns 255.255.255.0
+    :rtype: str
+    """
+    bits = 0xffffffff ^ (1 << 32 - mask) - 1
+    return socket.inet_ntoa(struct.pack('>I', bits))
+
+
+def is_ipv4_address(string_ip):
+    """
+    :rtype: bool
+    """
+    try:
+        socket.inet_aton(string_ip)
+    except socket.error:
+        return False
+    return True
+
+
+def get_no_proxy():
+    no_proxy = os.environ.get('no_proxy') or os.environ.get('NO_PROXY')
+
+    if no_proxy:
+        no_proxy = (host for host in no_proxy.replace(' ', '').split(',') if host and no_proxy)
+
+    # no_proxy in the proxies argument takes precedence
+    return no_proxy
+
+
+def bypass_proxy(host):
+    no_proxy = get_no_proxy()
+
+    if is_ipv4_address(host):
+        for proxy_ip in no_proxy:
+            if is_valid_cidr(proxy_ip):
+                if address_in_network(host, proxy_ip):
+                    return True
+            elif host == proxy_ip:
+                # If no_proxy ip was defined in plain IP notation instead of cidr notation &
+                # matches the IP of the index
+                return True
+    else:
+        for proxy_domain in no_proxy:
+            if host.endswith(proxy_domain) or host.split(':')[0].endswith(proxy_domain):
```

Comment:
we should use urlparse instead of string.split to get the hostname

should the string comparisons be case-insensitive?

Reply:
This is not required as the host variable already gets url parsed. The caller of `bypass_proxy(host)` already uses `_parse_url(url)` method.

Thus removing the second part of the code. Not required.

Adding the case-insensitivity to the code.